stan/math/prim/fun/lb_constrain.hpp

```diff
@@ -41,21 +49,104 @@ inline auto lb_constrain(const T& x, const L& lb) {
  * reference with the log absolute Jacobian determinant of the
  * transform.
  *
- * @tparam T Type of Matrix
- * @tparam L type of lower bound
- * @tparam S type of log probability
- * @param[in] x unconstrained Matrix input
+ * @tparam T Scalar.
+ * @tparam L Scalar.
+ * @param[in] x unconstrained input
  * @param[in] lb lower bound on output
  * @param[in,out] lp reference to log probability to increment
  * @return lower-bound constrained value corresponding to inputs
  */
-template <typename T, typename L>
+template <typename T, typename L, require_all_stan_scalar_t<T, L>* = nullptr,
+          require_all_not_st_var<T, L>* = nullptr>
 inline auto lb_constrain(const T& x, const L& lb, return_type_t<T, L>& lp) {
-  auto& x_ref = to_ref(x);
-  auto& lb_ref = to_ref(lb);
-  check_finite("lb_constrain", "lb", value_of(lb_ref));
-  lp += sum(x_ref);
-  return eval(add(exp(x_ref), lb_ref));
+  if (lb == NEGATIVE_INFTY) {
+    return identity_constrain(x, lb);
+  } else {
+    // check_less("lb_constrain", "lb", value_of(x), value_of(lb));
+    lp += x;
+    return add(exp(x), lb);
+  }
+}
+
+/**
+ * Specialization of `lb_constrain` to apply a scalar lower bound elementwise
+ *  to each input.
+ *
+ * @tparam T A type inheriting from `EigenBase`.
+ * @tparam L Scalar.
+ * @param[in] x unconstrained input
+ * @param[in] lb lower bound on output
+ * @return lower-bound constrained value corresponding to inputs
+ */
+template <typename T, typename L, require_eigen_t<T>* = nullptr,
+          require_stan_scalar_t<L>* = nullptr,
+          require_all_not_st_var<T, L>* = nullptr>
+inline auto lb_constrain(T&& x, L&& lb) {
+  return make_holder(
+      [](const auto& x, const auto& lb) {
+        return x.unaryExpr([lb](auto&& x) { return lb_constrain(x, lb); });
+      },
+      std::forward<T>(x), std::forward<L>(lb));
```

Comment:
[optional] I'd rather just evaluate expressions here because I'm pretty sure we're always assigning to l-values somewhere in generated C++. I think it does cost a malloc or something to make a holder? @t4c1 can confirm if they're free or not.

Reply:
Yeah good point

Reply:
If a holder is constructed from lvalues only it is virtually free. We need a malloc for each rvalue. But let me ask again: Do we even need holders here?